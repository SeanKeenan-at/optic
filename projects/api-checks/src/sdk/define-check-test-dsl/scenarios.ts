import { OpenAPIV3 } from 'openapi-types';
const HttpMethods = OpenAPIV3.HttpMethods;
export type BeforeAndAfter = [OpenAPIV3.Document, OpenAPIV3.Document];
export const copy = (obj: any) => JSON.parse(JSON.stringify(obj));

const defaultEmpty = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {},
});

const defaultWithoutSchema = (): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
          },
        },
      },
    },
  },
});

const defaultWithSchema = (
  schema: OpenAPIV3.SchemaObject
): OpenAPIV3.Document => ({
  openapi: '3.0.1',
  info: {
    version: '0.1.0',
    title: '',
  },
  paths: {
    '/example': {
      get: {
        responses: {
          '200': {
            description: '',
            content: {
              'application/json': {
                schema: schema,
              },
            },
          },
        },
      },
    },
  },
});

export function scenario(name: string) {
  return {
    responseSchema: {
      added: (schema: OpenAPIV3.SchemaObject): BeforeAndAfter => {
        return [defaultWithoutSchema(), defaultWithSchema(schema)];
      },
      changed: (
        beforeSchema: OpenAPIV3.SchemaObject,
        afterSchema: OpenAPIV3.SchemaObject
      ): BeforeAndAfter => {
        return [
          defaultWithSchema(beforeSchema),
          defaultWithSchema(afterSchema),
        ];
      },
    },
    operation: {
      requirement: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const spec: OpenAPIV3.Document = defaultEmpty();
        spec.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [spec, spec];
      },
      added: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        copied.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [defaultEmpty(), copied];
      },
      removed: (
        operation: OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const copied: OpenAPIV3.Document = defaultEmpty();
        copied.paths = {
          [pathPattern]: {
            [method]: operation,
          },
        };
        return [copied, defaultEmpty()];
      },
      changed: (
        operationBefore: OpenAPIV3.OperationObject,
        operationAfter: (
          before: OpenAPIV3.OperationObject
        ) => OpenAPIV3.OperationObject,
        method: OpenAPIV3.HttpMethods = HttpMethods.GET,
        pathPattern: string = '/example'
      ): BeforeAndAfter => {
        const before: OpenAPIV3.Document = defaultEmpty();
        before.paths = {
          [pathPattern]: {
            [method]: operationBefore,
          },
        };
        const after: OpenAPIV3.Document = defaultEmpty();
        after.paths = {
          [pathPattern]: {
            [method]: operationAfter(copy(operationBefore)),
          },
        };
        return [before, after];
      },
    },
  };
}
