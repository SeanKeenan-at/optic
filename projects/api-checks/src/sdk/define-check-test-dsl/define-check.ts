import { BeforeAndAfter } from './scenarios';
import { ApiCheckService } from '../api-check-service';
import { ApiChangeDsl } from '../api-change-dsl';
import { Result } from '../types';
import invariant from 'ts-invariant';

export interface IApiCheckDefinition {
  name: string;
  implementation?: (apiChangeDsl: ApiChangeDsl) => void;
  description?: string;
  validExamples: BeforeAndAfter[];
  invalidExamples: BeforeAndAfter[];
}

class ApiCheckImpl {
  public check: IApiCheckDefinition;

  constructor(name: string) {
    this.check = {
      name,
      validExamples: [],
      invalidExamples: [],
    };
  }

  description(description: string) {
    this.check.description = description;
    return this;
  }

  implementation(
    implementation: (apiChangeDsl: ApiChangeDsl) => void | Promise<void>
  ) {
    this.check.implementation = implementation;
    return this;
  }

  passingExample(beforeAndAfter: BeforeAndAfter) {
    this.check.validExamples = [...this.check.validExamples, beforeAndAfter];

    if (!this.check.implementation) test.skip('', () => 1);

    if (typeof jest !== 'undefined' && this.check.implementation) {
      // @ts-ignore
      const { test, expect } = global;

      const testName = `passing case ${this.check.validExamples.length}: ${beforeAndAfter[2]}`;

      test(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeTruthy();
      });
    }

    return this;
  }

  failingExample(beforeAndAfter: BeforeAndAfter) {
    this.check.invalidExamples = [
      ...this.check.invalidExamples,
      beforeAndAfter,
    ];

    if (!this.check.implementation) test.skip('', () => 1);

    if (typeof jest !== 'undefined' && this.check.implementation) {
      // @ts-ignore
      const { it, expect } = global;

      const testName = `failing case ${this.check.invalidExamples.length}: ${beforeAndAfter[2]}`;
      it(testName, async () => {
        const testResult = await this.testExample(beforeAndAfter);
        expect(testResult).toMatchSnapshot();
        expect(testResult.passed).toBeFalsy();
      });
    }

    return this;
  }

  private async testExample(beforeAndAfter: BeforeAndAfter) {
    invariant(this.check.implementation);
    const service = new ApiCheckService().useRulesBuildFrom(
      this.check.implementation
    );

    const [before, after] = beforeAndAfter;

    const runResults = await service.runRules(before, after, {});

    return {
      passed: !runResults.some((i) => !i.passed),
      results: runResults,
    };
  }
}

export function check(name: string) {
  return new ApiCheckImpl(name);
}
