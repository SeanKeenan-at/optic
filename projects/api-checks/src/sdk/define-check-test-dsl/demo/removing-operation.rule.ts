import { check } from '../define-check';
import { scenario } from '../scenarios';

const operationRemovalCheck = check('prevent operation removal')
  .description('this is a breaking change')
  .implementation(({ operations }) => {
    const { expect } = require('chai');
    operations.removed.must('not be removed unless marked deprecated', () => {
      expect.fail('operations can not be removed');
    });
  })
  .passingExample(
    scenario('updating description for GET /example').operation.changed(
      {
        operationId: '123',
        responses: {
          200: {
            description: 'abc',
          },
        },
      },
      (original) => {
        original.description = 'CHANGED IT';
        return original;
      }
    )
  )
  .failingExample(
    scenario('removing').operation.removed({
      responses: {},
    })
  );
