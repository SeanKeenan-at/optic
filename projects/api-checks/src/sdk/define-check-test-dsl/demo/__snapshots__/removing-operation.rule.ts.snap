// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` failing case 1 1`] = `
Object {
  "passed": false,
  "results": Array [
    Object {
      "change": Object {
        "changeType": "removed",
        "location": Object {
          "conceptualLocation": Object {
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
          ],
          "jsonPath": "/paths/~1example/get",
          "kind": "operation",
        },
        "removed": Object {
          "before": Object {
            "method": "get",
            "pathPattern": "/example",
          },
        },
      },
      "condition": "not be removed unless marked deprecated",
      "docsLink": undefined,
      "error": "operations can not be removed",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "removed operation: GET /example",
    },
  ],
}
`;

exports[` failing case 2 1`] = `
Object {
  "passed": true,
  "results": Array [],
}
`;

exports[` passing case 1 1`] = `
Object {
  "passed": true,
  "results": Array [],
}
`;
